phosphor-auditlog/auditLogParser.cpp

```diff
@@ -0,0 +1,463 @@
+#include "auditLogParser.hpp"
+
+#include "auditLogMgr.hpp"
+
+#include <auparse.h>
+#include <libaudit.h>
+
+#include <nlohmann/json.hpp>
+#include <phosphor-logging/lg2.hpp>
+#include <xyz/openbmc_project/Common/File/error.hpp>
+#include <xyz/openbmc_project/Common/error.hpp>
+
+#include <cstring>
+#include <filesystem>
+#include <list>
+#include <map>
+#include <string>
+
+namespace phosphor
+{
+namespace auditlog
+{
+
+/**
+ * @brief Moves parser to point to next event
+ *
+ * @return false when no more events exist, or on error
+ */
+bool ALParser::auditNextEvent()
+{
+    bool haveEvent = false;
+    int rc;
+
+    rc = auparse_next_event(au);
+    switch (rc)
+    {
+        case 1:
+            /* Success, pointing to next event */
+            haveEvent = true;
+            break;
+        case 0:
+            /* No more events */
+            haveEvent = false;
+            break;
+        case -1:
+        default:
+            /* Failure */
+            lg2::error("Failed to parse next event");
+            haveEvent = false;
+            break;
+    }
+
+    return haveEvent;
+}
+
+/**
+ * @brief Parses next event and each of its records into JSON format
+ * @details Writes the audit log events into the parsedFile.
+ * Parsed into format for bmcweb to read the parsedFile.
+ */
+void ALParser::parseEvent()
+{
+    unsigned int nRecords = auparse_get_num_records(au);
+
+    // The event itself is a record. It may be the only one.
+    parseRecord();
+
+    /* Handle any additional records for this event */
+    for (unsigned int iter = 1; iter < nRecords; iter++)
+    {
+        auto rc = auparse_next_record(au);
+
+        switch (rc)
+        {
+            case 1:
+            {
+                /* Success finding record, parse it! */
+                parseRecord();
+            }
+            break;
+            case 0:
+                /* No more records, something is confused! */
+                lg2::error(
+                    "Record count ({NRECS}) and records out of sync ({ITER})",
+                    "NRECS", nRecords, "ITER", iter);
+                throw sdbusplus::xyz::openbmc_project::Common::Error::
+                    InternalFailure();
+
+                break;
+            case -1:
+            default:
+                /* Error */
+                lg2::error("Failed on record number={ITER}", "ITER", iter);
+                throw sdbusplus::xyz::openbmc_project::Common::Error::
+                    InternalFailure();
+                break;
+        }
+    }
+}
+
+/**
+ * @brief Parses general audit entry into JSON format
+ * @details Used with audit entries without specific handling. Text of audit
+ * log message is written as-is.
+ */
+void ALParser::fillAuditEntry(nlohmann::json& parsedEntry)
+{
+    parsedEntry["MessageId"] = "OpenBMC.0.5.AuditLogEntry";
```

Comment:
May need to see what Gunnar thinks, but usually you'd want anything Redfish specific in bmcweb, and this code just returns the 'raw' data and bmcweb can deal with formatting it correctly for Redfish.  That way it will also work with some new future protocol.

Reply:
I will check with Gunnar on this. I modeled this after the EventLog format. The bmcweb handling of those entries it looks like the message registry id is included in the entry. (See [fillEventLogEntryJson](https://github.com/openbmc/bmcweb/blob/a88942019fdd3d8fc366999f7c178f3e1c18b2fe/redfish-core/lib/log_services.hpp#L1479))

Reply:
I am with Matt, I am surprised that Journal EventLog would have the MessageId. I think it should be contained to bmcweb   
https://github.com/openbmc/bmcweb/blob/a88942019fdd3d8fc366999f7c178f3e1c18b2fe/redfish-core/lib/log_services.hpp#L1504

Are there other ways to determine if this is an OpenBMC.0.5.AuditLogEntry or a different AuditLog? 
If you see some field, then assume it is an OpenBMC.0.5.AuditLogEntry ?

Reply:
That said, I don't care that much. 
I could live with this until this is all done upstream

Reply:
For my original prototype I was separating out the audit message type into the first index of the MessageArgs for both message types. But that was then pushing down understanding of the Linux audit message types into bmcweb which I didn't think would be acceptable especially upstream. That is why I then moved to the message registry approach since I had seen precedent of those outside of bmcweb.

Other possibilities:

- Key off the number of MessageArgs. One MessageArg, vs 8. Any other number would be an error.
- Define values that are understood by phosphor-auditlog and bmcweb to distinguish the two types. (An enum added to the phosphor-dbus-interface?)

Reply:
I'm leaning towards leaving this as-is for now and addressing it upstream. @spinler are you okay with that?

Reply:
sure.

Reply:
PRs created to address this comment: https://github.com/ibm-openbmc/phosphor-logging/pull/64 and https://github.com/ibm-openbmc/bmcweb/pull/1020